Scripting the creation of Octopus API keys (#409)
OctopusDeploy/blog: blog/2021-02/apikey-creation-automation/index.md

---
title: "Scripting the creation of Octopus API keys"

> description: Using PowerShell to simulate Octopus login and API key creation.

Reviewer: ```suggestion
description: Learn how to script the creation of an API key for use with the Octopus REST API, and how to use browser developer tools to explore the Octopus API in action.
```

author: phil.stephenson@octopus.com
visibility: public
published: 2999-01-01-1400
tags:
 - DevOps
 - PowerShell
 - API
---

Advanced users of Octopus will already be familiar with the robust, feature-rich [Octopus REST API](https://octopus.com/docs/octopus-rest-api). You may have used it to automate a unique process that isn't available in the Octopus Web UI, for example. Before you can interact with the API however, you need to create an Octopus [API key](https://octopus.com/docs/octopus-rest-api/how-to-create-an-api-key). These steps require a human though, and we know some of you would prefer to automate the process. 

In this blog post, I walk through scripting the creation of an API key for use with the Octopus REST API.

TL;DR - Just want to see the final script? Check out [my GitHub gist](https://gist.github.com/pstephenson02/3cf2dc3b9d68db28722ad568c9eb49eb).

> ## Swagger

Reviewer: ```suggestion
## The Swagger API docs
```

![Swagger ApiKeys resource screenshot](find-apikey.png "width=500")

The first row under ApiKeys: `POST /users/{userId}/apikeys` allows us to create a new API key for the specified user. API keys in Octopus are associated with an Octopus user, and the keys inherit the permissions assigned to that user.

:::hint
It's best practice when provisioning new API keys to set up [Octopus Service Accounts](https://octopus.com/docs/security/users-and-teams/service-accounts) dedicated for specific functions or integrations.
:::

Let's assume we know our `{userId}` and use a quick one line PowerShell cmdlet to create an API key:

```pwsh
> Invoke-RestMethod -Method Post -Uri https://samples.octopus.app/api/users/Users-561/apikeys -Headers @{'X-Octopus-ApiKey'='API-XXXX...'} -Body (@{'Purpose'='Just blogging'} | ConvertTo-Json)

Id      : apikeys-I6D74k9rh7eyoqXDlqJCvlsVgU
Purpose : Just blogging
UserId  : Users-561
ApiKey  : API-XXXX...
Created : 2/18/2021 2:25:49 PM
Expires :
Links   : @{Self=/api/users/Users-561/apikeys/apikeys-I6D74k9rh7eyoqXDlqJCvlsVgU}
```

You might notice in this example, that we already have an API to create a new API key.

> ## Create an API key when you don't already have one

Reviewer: ```suggestion
## Creating an API key when you don't already have one
```

Imagine you're writing automation to provision your Octopus Server itself. You've written scripts using the [Octopus Deploy Chocolatey Package](https://chocolatey.org/packages/OctopusDeploy/), the [Octopus.Server.exe command line tool](https://octopus.com/docs/octopus-rest-api/octopus.server.exe-command-line), or even the new [Octopus Deploy Terraform Provider](https://octopus.com/blog/octopusdeploy-terraform-provider). You don't want to break off in the middle of your provisioning automation because you need to login to your Octopus Server, create an API key, then manually plug the key into the rest of your automation.

> When we create new users in Octopus, those users don't have API keys, but the Octopus Web Portal lets us create the user is logged in. And if the browser can do it, then so can we. Here's what to do:

Reviewer 2: ```suggestion
When we create new users in Octopus, those users don't have API keys, but the Octopus Web Portal lets us create APK keys when the user is logged in. And if the browser can do it, then so can we. Here's what to do:
```

Reviewer 2: How is this @pstephenson02 ?

Author: Looks great. Thank you 👍

1. Simulate the browser login with a username and password.
2. Retrieve any necessary cookies sent back to us from Octopus Server.
3. Use our cookies to make the same request the browser does when creating a user's first API key.

When we first [install Octopus Server](https://octopus.com/docs/installation#install-octopus), it asks us to create a `Local System Account` or `Custom Domain Account`. For simplicity, let's assume you have a Local System Account. You can also create one using the `Octopus.Server.exe` CLI's [`admin` command](https://octopus.com/docs/octopus-rest-api/octopus.server.exe-command-line/admin).

> ## Chrome DevTools

Reviewer: ```suggestion
## Inspecting browser activity with Chrome Dev Tools
```

> ## Chrome DevTools

Reviewer: ```suggestion
## Inspecting browser activity with Chrome DevTools
```

Reviewer 2: This change has been made

> We need to inspect what's happening in the browser when we log into our Octopus Web Portal. Let's use [Chrome's excellent built-in DevTools](https://developers.google.com/web/tools/chrome-devtools). 

Reviewer: ```suggestion
We need to inspect what's happening in the browser when we log into our Octopus Web Portal. Let's use [Chrome's built-in DevTools](https://developers.google.com/web/tools/chrome-devtools). 
```

Links   : @{Self=/api/users/Users-561/apikeys/apikeys-I6D74k9rh7eyoqXDlqJCvlsVgU}
```

You might notice in this example, that we already have an API to create a new API key.

## Create an API key when you don't already have one

Imagine you're writing automation to provision your Octopus Server itself. You've written scripts using the [Octopus Deploy Chocolatey Package](https://chocolatey.org/packages/OctopusDeploy/), the [Octopus.Server.exe command line tool](https://octopus.com/docs/octopus-rest-api/octopus.server.exe-command-line), or even the new [Octopus Deploy Terraform Provider](https://octopus.com/blog/octopusdeploy-terraform-provider). You don't want to break off in the middle of your provisioning automation because you need to login to your Octopus Server, create an API key, then manually plug the key into the rest of your automation.

> When we create new users in Octopus, those users don't have API keys, but the Octopus Web Portal lets us create the user is logged in. And if the browser can do it, then so can we. Here's what to do:

Author: @wordlee thanks for your edits here. This is the only one I'm a little unclear on. Specifically the end of the first sentence. It seems like maybe one or two many words were removed?

Reviewer 2: Thanks @pstephenson02. Yeah, that's not quite right. I'll fix that 👍

1. Simulate the browser login with a username and password.
2. Retrieve any necessary cookies sent back to us from Octopus Server.
3. Use our cookies to make the same request the browser does when creating a user's first API key.

When we first [install Octopus Server](https://octopus.com/docs/installation#install-octopus), it asks us to create a `Local System Account` or `Custom Domain Account`. For simplicity, let's assume you have a Local System Account. You can also create one using the `Octopus.Server.exe` CLI's [`admin` command](https://octopus.com/docs/octopus-rest-api/octopus.server.exe-command-line/admin).

## Chrome DevTools

We need to inspect what's happening in the browser when we log into our Octopus Web Portal. Let's use [Chrome's excellent built-in DevTools](https://developers.google.com/web/tools/chrome-devtools). 

1. Navigate to your Octopus Web Portal login page.
2. Either right click on the page and select `Inspect`, or use the hot keys `Ctrl-Shift-i` to open DevTools. 
3. Select the `Network` tab shown in the screenshot.

> 4. After entering your username and password, click `Sign In`.

Author: This works so much better in a list! 👍
